pennylane/workflow/execution.py

```diff
@@ -259,13 +260,10 @@ def _make_inner_execute(
 
     if isinstance(device, qml.Device):
         device_execution = set_shots(device, override_shots)(device.batch_execute)
-
     else:
         device_execution = partial(device.execute, execution_config=execution_config)
 
-    cached_device_execution = qml.workflow.cache_execute(
-        device_execution, cache, return_tuple=False
-    )
+    cached_device_execution = _apply_cache_transform(fn=device_execution, cache=cache)

```

Comment:
I think at this point we can just directly use `_cache_transform` inside `inner_execute`.

```
new_batch, postprocessing = _cache_transform(tapes)
device_results = device_execution(new_batch)
postprocessing(results)
```


I'd like to move toward a place where we have an "Inner transform program":

```
program = TransformProgram()
if expand_fn:
    program.add_transform(expand_fn_transform(expand_fn))
if numpy_only:
    program.add_transform(convert_to_numpy_parameters)
program.add_transform(_cache_transform)

new_batch, postprocessing = program(tapes)
device_results = device_execution(new_batch)
postprocessing(device_results)
```

We would just need to convert `convert_to_numpy_parameters` to a transform and duplicate the `expand_fn_transform` code from `workflow/construct_batch.py`. So eventually we'll get there.

Reply:
@albi3ro I think that suggestion makes a lot of sense! That was actually my original approach but I ended up going with `_apply_cache_transform()` to reuse the same code when calculating `"device"` gradients in `execute()` (see `execution.py` line ~800).

As such, do you think it makes sense to adopt the `TransformProgram` style in `_make_inner_execute()` but use `_apply_cache_transform()` in the device gradient `else` branch? Or should we use it in both places? 

We could also potentially keep `_apply_cache_transform()` for now and switch to using `TransformProgram` when we get around to rewriting `convert_to_numpy_parameters()` as a transform.

Reply:
So the `"device"` gradients logic is now in low-maintainence mode, and at some point I want to just relegate all that logic into `tensorflow_autograph.py` and let it deal with itself. So the philosophy with that code is "keep it working with the minimal amount of thought and effort".

But for `_make_inner_execute`, I think it is worth moving it in the direction we want it to be, even if that is slowly.

But I'm good with both approaches.  This file will 🤞 improve someday.